### PR TITLE
Added support for uploading and downloading binary data

### PR DESCRIPTION
**Pull request by sandves**
Adds the following methods to the `IHttpClient` interface:

```typescript
postFormAsync<TResponse>(
    url: string,
    form: FormData,
    onProgress?: (percentage: number, event: ProgressEvent<XMLHttpRequestEventTarget>) => void
): Promise<HttpResponse<TResponse>>;
```

```typescript
getBlobAsync<TExpectedErrorResponse>(url: string): Promise<File>;
```

They will be used to upload and download binary data like files and attachments.

### CODE REVIEW

File: src/http/HttpClient/index.ts [hunk 1]
```diff
@@ -169,6 +169,102 @@ export default class HttpClient implements IHttpClient {
...
+            xhr.send(form);
+        });
+    }
+
+    async getBlobAsync<TExpectedErrorResponse>(
+        url: string,
+        init?: RequestInit | null
+    ): Promise<File> {
+        if (!init) {
+            {
```
Comment (MartinForre) [inline src/http/HttpClient/index.ts]: This does nothing?

File: package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@equinor/fusion",
-    "version": "0.4.60",
+    "version": "0.5.00",
```
Comment (eskil0312) [inline package.json]: Should we instead do the versioning in the master branch?

File: src/http/HttpClient/index.ts [hunk 3]
```diff
@@ -169,6 +169,102 @@ export default class HttpClient implements IHttpClient {
...
+        init = ensureRequestInit(init, init => ({
+            ...init,
+            method: 'GET',
+        }));
+
+        const response = await this.performFetchAsync<TExpectedErrorResponse>(url, init);
+        const blob = await response.blob();
+
+        const contentDisposition = response.headers.get('Content-Disposition');
+        let filename;
```
Comment (eskil0312) [inline src/http/HttpClient/index.ts]: Could this be done using Array.Prototype.Find ? Maybe something like this

 
```suggestion
    if (contentDisposition === null) {
        throw new Error('Cannot download file without filename');
    }
    const parts = contentDisposition.split(';');
    const fileNamePart = parts.find(part =>  part.indexOf('filename=') !== -1)
    const filename = fileNamePart.split('=')[1];


    return new File([blob], filename);
```

Doing it this way we avoid nested if's and variables

File: src/http/HttpClient/index.ts [hunk 4]
```diff
@@ -169,6 +169,102 @@ export default class HttpClient implements IHttpClient {
...
+                    }
+                });
+            }
+
+            xhr.addEventListener('load', () => {
+                const headerLines = xhr.getAllResponseHeaders();
+                const headers = headerLines.trim().split(/[\r\n]+/);
+
+                var headerMap = new Headers();
+                headers.forEach(function(line) {
```
Comment (eskil0312) [inline src/http/HttpClient/index.ts]: Arrow function? Could this be done with a reducer?

File: src/http/HttpClient/index.ts (hunk 3 above)
Comment (eskil0312) [reply src/http/HttpClient/index.ts]: The suggested change didn't work over multiple lines, so don't commit from here :D

Comment (sandves) [reply src/http/HttpClient/index.ts]: Good point. Extraxted the file name logic to a helper method.

File: src/http/HttpClient/index.ts (hunk 4 above)
Comment (sandves) [reply src/http/HttpClient/index.ts]: Like this?

```javascript
const headerMap = headers.reduce((headerMap, line) => {
    const parts = line.split(': ');
    const header = parts.shift();
    const value = parts.join(': ');
    if (header) headerMap.append(header, value);
    return headerMap;
}, new Headers());
```

I'm not sure which one is more readable, but we can go with reducer if that is the convention for doing operations like this.

Comment (eskil0312) [reply src/http/HttpClient/index.ts]: I guess they are pretty much equally readable. Go with your choice

Comment (eskil0312) [reply src/http/HttpClient/index.ts]: But by using a reducer we don't have to deal with vars

**Pull request merged**